### PR TITLE
Nazrhom/snapshot exporter

### PR DESCRIPTION
**Pull request by nazrhom**

### CODE REVIEW

**Review by nrutledge — APPROVED** [review]
I added some comments about little nitpicky things, but otherwise looks good.

File: components/cardano-node/dowload-mithril-snapshot.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ROOT_DIR="$COMPONENT_DIR/.."
+
+# shellcheck source=/dev/null
+source "${ROOT_DIR}/configurations/mithril-configs/${NETWORK}.env"
+
+rm -rf "$ROOT_DIR/db"
+
+mithril-client snapshot download latest
+
+mv "$ROOT_DIR/db/*" "$ROOT_DIR/mithril-snapshot"
```
Comment (aciceri) [inline components/cardano-node/dowload-mithril-snapshot.sh:17]:
```suggestion
mv "$ROOT_DIR/db" "$ROOT_DIR/mithril-snapshot"
```

Did you mean this?

**Review by aciceri — COMMENTED** [review]
When I run `npm test` I get a lot of errors like this:
```
test: Error: ENOENT: no such file or directory, lstat '/home/ccr/projects/mlabs-haskell/cardanow/test.txt'
```

File: snapshot.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+export class SnapshotExporter {
+  readonly config: SnapshotConfig
+  readonly cadence: number // Execution cadence in milliseconds. The first run will wait for `cadence` ms before executing
+  readonly maxAttempts: number // Maximum number of attempts
+  attempt: number // Current attempt
+  
+  constructor (config: SnapshotConfig, cadence?: number, maxAttempts?: number) {
+    this.config = config
+    this.maxAttempts = maxAttempts || 200
```
[the remark this answers is not in the document]
Comment (nazrhom) [reply snapshot.ts:48]: That was what I had initially, but it either forces me to duplicate the config object type, or have to add `!` at every use within the class. I thought leaving the optional params as optional constructor args was a good compromise.
In this case you can simply omit them if you don't want to change them, no need to pass in `null`s or `undefined`s at the call site.

File: snapshot.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    this.cadence = cadence || 600000
+    this.attempt = 0
+  }
+
+  private runWithDelay: Action<string> = async () => {
+    this.attempt++
+    console.log(`${this.config.name}: Running attempt #${this.attempt}`)
```
[the remark this answers is not in the document]
Comment (nazrhom) [reply snapshot.ts:55]: Yeah, this is intended, I thought the delay is part of the attempt after all, happy to go either way here though

File: components/cardano-node/dowload-mithril-snapshot.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+set -o nounset
+set -o errexit
+set -o pipefail
+
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+COMPONENT_DIR="$SCRIPT_DIR/.."
+ROOT_DIR="$COMPONENT_DIR/.."
```
[the remark this answers is not in the document]
Comment (nazrhom) [reply components/cardano-node/dowload-mithril-snapshot.sh:8]: what kind of error did you get here? is it maybe not running on bash?

File: components/cardano-node/dowload-mithril-snapshot.sh (hunk 1 above)
Comment (nazrhom) [reply components/cardano-node/dowload-mithril-snapshot.sh:17]: The one you suggested creates a `db` folder inside `mithril-snapshot`, I only want to move the contents of the db folder

File: snapshot.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  private handleSuccess: Action<string> = async () => {
+    const sha: string = await this.pkg()
+    console.log(`${this.config.name}: Done`)
+    return sha
+  }
+
+  private handleFailure: Action<string> = async () => {
+    if (this.attempt < this.maxAttempts) {
```
[the remark this answers is not in the document]
Comment (nazrhom) [reply snapshot.ts:82]: This is just a NIT, but I slightly prefer the original version, as the explicit `if/else` makes it easier to reason about flow. In terms of `reject` vs `throw` it should be equivalent (if not slightly better to use `reject` for some weird corner cases, e.g. https://stackoverflow.com/questions/33445415/javascript-promises-reject-vs-throw). I don't mind too much either way though, if you prefer we stick to throwing I can switch this and follow that in the future.

**Pull request merged**